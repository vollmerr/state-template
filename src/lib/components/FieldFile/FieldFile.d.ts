import * as React from 'react';

export interface FieldFileProps extends HTMLInputElement {
  /** Accessible indicator of related information */
  'aria-describedby'?: string;

  /** Accessible indicator for errors existing */
  'aria-invalid'?: string;

  /** Text for upload button */
  btnText?: string | React.ReactNode;

  /** What source to use for capturing image or video data */
  capture?: string;

  /** Ref to attach to input */
  inputRef?: React.Ref<HTMLInputElement>;

  /** Label for the field */
  label: string;

  /** Name of field */
  name: string;

  /** Called when the field is blurred */
  onBlur?: (event: FocusEvent, value?: string) => void;

  /** Called when the field changes */
  onChange?: (event: Event, value?: string) => void;

  /** Called when the field is focused */
  onFocus?: (event: FocusEvent, nodeType: string) => void;
}

declare class FieldFile extends React.Component<FieldFileProps, {}> { }

export default FieldFile;
