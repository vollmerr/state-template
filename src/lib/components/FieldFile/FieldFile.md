```jsx
const { file } = require('../../index');

const downloadButton = {
  text: 'download `File`',
  onClick: (values) => {
    if (values.file) {
      file.saveFile(values.file, values.file.name);
    }
  },
};

<ExampleForm form={'file'} customButton={downloadButton}>
  <FieldFile
    name={'file'}
    label={'File'}
    helpText={'This is the `helpText` content...'}
    onChange={(value) => { console.log('changing...', value) }}
    onBlur={(value) => { console.log('blurring...', value) }}
    onFocus={(value) => { console.log('focusing...', value) }}
    accept={'.jpg, .jpeg, .png'}
    placeholder={'Custom Placeholder...'}
    btnText={'Custom Button Text'}
  />
  <FieldFile 
    required
    name={'requiredFile'}
    label={'Required File'}
  />
  <FieldFile 
    disabled
    name={'disabledFile'}
    label={'Disabled File'}
  />
  <FieldFile
    multiple
    name={'multipleFile'}
    label={'Multiple Files'}
  />
</ExampleForm>
```
