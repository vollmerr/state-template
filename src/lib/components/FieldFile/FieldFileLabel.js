import React from 'react';
import T from 'prop-types';

/**
 * Label to render above field. Actual <label> element is inside the file picker,
 * used to interact with the input...
 */
const FieldFileLabel = ({ label, required }) => {
  if (!label) {
    return null;
  }

  return (
    <div data-test={'label'} className={'field__file-label'}>
      {
        required
        && <span data-test={'required'} className={'required-label'} aria-hidden>* </span>
      }

      {label}
    </div>
  );
};

FieldFileLabel.propTypes = {
  label: T.string,
  required: T.bool,
};

FieldFileLabel.defaultProps = {
  label: null,
  required: false,
};

export default FieldFileLabel;
