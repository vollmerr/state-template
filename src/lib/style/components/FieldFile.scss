.field__file-label {
  font-weight: 700;
  margin-bottom: .3rem;
}

.field__file {
  position: relative;

  input:not(:disabled) {
    &:active,
    &:focus {
      + .field__file-text {
        color: #495057;
        background-color: #fff;
        border-color: #80bdff;
        outline: 0;
        box-shadow: 0 0 0 0.2rem rgba(0,123,255,.25);
      }
    }
  }

  .field__file-text {
    margin: 0;
    cursor: pointer;
  }

  .field__file-button {
    position: absolute;
    padding: 0 15px;
    top: 0;
    right: 0;
    bottom: 0;
    pointer-events: none;
    background: $btn-primary-bg;
    color: $btn-primary-color;
    line-height: 38px;
  }

  input {
    z-index: -1;
    position: absolute;
    opacity: 0;
  }
}

.disabled {
  .field__file {
    .field__file-text {
      color: $disabled-fg;
      border-color: $disabled-accent;
      background: $disabled-bg;
      cursor: default;
    }

    .field__file-button {
      display: none;
    }
  }
}

