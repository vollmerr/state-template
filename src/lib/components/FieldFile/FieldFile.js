import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

import { isEmptyFile } from '../../utils/validate';
import { withField } from '../Field';

import FieldFileLabel from './FieldFileLabel';

// file attachment field that applies state-template styling
export class FieldFile extends React.PureComponent {
  inputRef = React.createRef()

  componentWillReceiveProps(nextProps) {
    const { value } = nextProps;
    // handle resetting the form
    if (!value) {
      this.inputRef.current.value = '';
    }
  }

  eventHandler = func => () => {
    const { multiple } = this.props;
    const { files } = this.inputRef.current;
    const value = multiple ? files : files[0];
    return func(value || null);
  }

  getFileText = () => {
    const { placeholder } = this.props;
    const { current } = this.inputRef;

    if (current) {
      const { files } = current;
      // multiple files
      if (files && files.length > 1) {
        return `${files.length} files`;
      }
      // single file
      if (files[0]) {
        return files[0].name;
      }
    }
    // no file
    return placeholder || 'No File Selected';
  }

  render() {
    const {
      name,
      className,
      value, // will be passed as `files`, do not pass value
      onChange,
      onBlur,
      onFocus,
      btnText,
      label,
      ...rest
    } = this.props;

    const cn = classNames([
      'field__file',
      className,
    ]);

    return (
      <div data-test={'field__file'} className={cn}>
        <input
          type={'file'}
          onChange={this.eventHandler(onChange)}
          onBlur={this.eventHandler(onBlur)}
          onFocus={this.eventHandler(onFocus)}
          files={value}
          ref={this.inputRef}
          {...rest}
        />

        <label className={'form-control field__file-text'} htmlFor={name}>
          {this.getFileText()}
        </label>

        <div className={'field__file-button'}>
          {btnText}
        </div>
      </div>
    );
  }
}

FieldFile.propTypes = {
  /** One or more unique file type specifiers describing file types to allow */
  accept: T.string,

  /** Accessible indicator of related information */
  'aria-describedby': T.string,

  /** Accessible indicator for errors existing */
  'aria-invalid': T.string,

  /** Text for upload button */
  btnText: T.oneOfType([
    T.string,
    T.node,
  ]),

  /** What source to use for capturing image or video data */
  capture: T.string,

  /** Class names to attach to the field */
  className: T.string,

  /** Disable the input */
  disabled: T.bool,

  /** Id of field */
  id: T.string,

  /** aria-label for the field */
  label: T.string.isRequired,

  /** indicates that the user may choose more than one file */
  multiple: T.bool,

  /** Name of field */
  name: T.string.isRequired,

  /** Called when radio button is blurred */
  onBlur: T.func,

  /** Called when radio button changes */
  onChange: T.func,

  /** Called when radio button is focused */
  onFocus: T.func,

  /** Displayed when no file is selected */
  placeholder: T.string,

  /** Value of the field */
  value: T.any,
};

FieldFile.defaultProps = {
  accept: null,
  'aria-describedby': null,
  'aria-invalid': 'false',
  btnText: 'Choose File',
  capture: null,
  className: null,
  disabled: false,
  id: null,
  multiple: false,
  onBlur: null,
  onChange: null,
  onFocus: null,
  placeholder: null,
  value: null,
};

const options = {
  renderLabel: FieldFileLabel,
};

export default withField(isEmptyFile)(FieldFile, options);
