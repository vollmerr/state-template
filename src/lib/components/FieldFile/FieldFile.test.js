import React from 'react';
import { shallow } from 'enzyme';

import { FieldFile } from './FieldFile';

const props = {
  accept: '.jpg, .jpeg, .png',
  'aria-describedby': 'aria-desc-id',
  'aria-invalid': 'false',
  btnText: 'upload-button-text',
  className: 'test-classname',
  disabled: false,
  id: 'test-id',
  label: 'test-label',
  multiple: false,
  name: 'test-name',
  onBlur: jest.fn(),
  onChange: jest.fn(),
  onFocus: jest.fn(),
  placeholder: 'test-placeholder',
  value: 'test-value',
};

let wrapper;
describe('FieldFile', () => {
  beforeEach(() => {
    wrapper = shallow(<FieldFile {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
