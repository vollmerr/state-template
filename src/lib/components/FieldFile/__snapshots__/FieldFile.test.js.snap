// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldFile should render correctly 1`] = `
<div
  className="field__file test-classname"
  data-test="field__file"
>
  <input
    accept=".jpg, .jpeg, .png"
    aria-describedby="aria-desc-id test-name--filename"
    aria-invalid="false"
    aria-label="test-label"
    capture={null}
    disabled={false}
    files="test-value"
    id="test-id"
    multiple={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="test-placeholder"
    type="file"
  />
  <label
    aria-hidden={true}
    className="form-control field__file-text"
    htmlFor="test-name"
  >
    test-placeholder
  </label>
  <span
    hidden={true}
    id="test-name--filename"
  >
    test-placeholder
  </span>
  <div
    className="field__file-button"
  >
    upload-button-text
  </div>
</div>
`;
